### PR TITLE
team-level std algos: part 9

### PR DESCRIPTION
**Pull request by fnrizzi**
Co-authored-by: Cezary Skrzyński
Co-authored-by: Jakub Strzebonski

## Content 

Team-level implementation and tests for:

- `Kokkos_Unique.hpp`
- `Kokkos_AdjacentDifference.hpp`
- `Kokkos_Reduce.hpp`
- `Kokkos_TransformReduce.hpp`

## Technical comments

- the change `class` -> `typename` and the enable if defaulted in the template parameters comes from [this request](https://github.com/kokkos/kokkos/pull/6200#issuecomment-1626406816)

- the code duplication is intentional and agreed upon after the long discussion in [here](https://github.com/kokkos/kokkos/pull/6200)

- tests are run for both contiguous views and strided views

## Other comments

- all team-level PRs are tracked in: https://github.com/kokkos/kokkos/issues/6352

- this comes from breaking apart https://github.com/kokkos/kokkos/pull/5150

## Possible conflicts

Implementation-wise, this has not conflict with other team-level PRs so it could be merged independently. 
However, a trivial conflict in the CMakeLists inside the unit test occurs if another of the team-level PRs is merged before this because each team-level PR adds test using a specific "label". Resolving this is trivial. 

## Associated wiki PRs

- `Kokkos_Unique.hpp`: see [wiki PR](https://github.com/kokkos/kokkos-core-wiki/pull/434)
- `Kokkos_AdjacentDifference.hpp`, `Kokkos_Reduce.hpp`, `Kokkos_TransformReduce.hpp`: see [wiki PR](https://github.com/kokkos/kokkos-core-wiki/pull/440)

IMPORTANT: wiki PRs are not complete yet but 99% there 

The issue tracking all *wiki PRs* is [here](https://github.com/kokkos/kokkos-core-wiki/issues/379)

### CODE REVIEW

File: algorithms/unit_tests/TestStdAlgorithmsTeamReduce.cpp [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+  auto reduceResultsView_h = create_host_space_copy(reduceResultsView);
+
+  for (std::size_t i = 0; i < dataView.extent(0); ++i) {
+    auto rowFrom = Kokkos::subview(dataViewBeforeOp_h, i, Kokkos::ALL());
+
+    const auto rowFromBegin = KE::cbegin(rowFrom);
+    const auto rowFromEnd   = KE::cend(rowFrom);
+    const auto initVal      = reductionInitValuesView_h(i);
+
+#if defined(__GNUC__) && __GNUC__ == 8
```
Comment (Rombur) [inline algorithms/unit_tests/TestStdAlgorithmsTeamReduce.cpp:179]: Can  you add a comment why you are doing this.

**Comment by fnrizzi** [comment]
need to fix the single in Unique

**Comment by fnrizzi** [comment]
will rebase asa the current CI finishes so i don't waste a cycle

**Comment by fnrizzi** [comment]
Retest this please

**Comment by fnrizzi** [comment]
will remove the old impl in Unique once i see CI passing

**Review by crtrott — APPROVED** [review]

**Review by cz4rs — APPROVED** [review]

**Comment by fnrizzi** [comment]
CI failure for gcc only build unrelated

**Pull request merged**